### PR TITLE
feat(blurhash): Non-blocking blurhash decoding

### PR DESCRIPTION
**Pull request by ferferga**
This PR makes the decoding of blurhash a non-blocking operation, as it runs in another thread now:

### Done
* Fully written in TypeScript
* Worker works and pixels are returned to the main thread correctly

### Doesn't work:
* Putting the imagedata into the canvas doesn't work for some reason. Maybe because it's inside the worker event handler. Removing all the elements that are set on top of the canvas still shows that the canvas is empty (so we can discard that the canvas is rendered after the images has been loaded)

### CODE REVIEW

File: components/BlurhashCanvas.vue [hunk 1]
```diff
@@ -4,7 +4,8 @@
 
 <script lang="ts">
 import Vue from 'vue';
-import { decode } from 'blurhash';
+import Worker from 'worker-loader!./blurhash.worker';
+const worker = new Worker();
```
Comment (MrTimscampi) [inline components/BlurhashCanvas.vue:8]: This will create a new worker for each canvas (Since each canvas is a new component with its own lifecycle).

A better way to do this would be to have a generic web worker shared by the entire app (Created through a plugin, for example).

Generally, web workers should be reused as much as possible since, while not the most expensive thing around, they're still not instantaneous.

File: components/BlurhashCanvas.vue [hunk 2]
```diff
@@ -37,15 +38,21 @@ export default Vue.extend({
   },
   methods: {
     draw() {
-      const pixels = decode(this.hash, this.width, this.height);
-      if (pixels) {
-        const ctx = (this.$refs.canvas as HTMLCanvasElement).getContext('2d');
-        const imageData = ctx?.createImageData(this.width, this.height);
-        if (imageData) {
+      const bhash_data = {
```
Comment (MrTimscampi) [inline components/BlurhashCanvas.vue:41]: No need to assign this since it's used only once and right on the next line, imo.

File: components/BlurhashCanvas.vue [hunk 3]
```diff
@@ -37,15 +38,21 @@ export default Vue.extend({
...
+        hash: this.hash,
+        width: this.width,
+        height: this.height
+      };
+      worker.postMessage(bhash_data);
+      const ctx = (this.$refs.canvas as HTMLCanvasElement).getContext('2d');
+      const imageData = ctx?.createImageData(this.width, this.height);
+      worker.onmessage = ({ data }) => {
```
Comment (MrTimscampi) [inline components/BlurhashCanvas.vue:49]: I think a better way to do this would be to set a watcher on the hash and size, which would send the message to the worker, then define the onmessage in "created" (since it's an event, it seems to be what you're supposed to do) and have it call upon the `draw` method with the pixel array.

File: components/blurhash.worker.ts [hunk 4]
```diff
@@ -0,0 +1,12 @@
+import { decode, isBlurhashValid } from 'blurhash';
+
+const ctx: Worker = self as any;
```
Comment (MrTimscampi) [inline components/blurhash.worker.ts:3]: I'm not a fan of this.

Firstly, this is not a component, so it shouldn't live in the Components folder (That's purely for Vue components).

Secondly, this pretty much disable types for anything regarding the worker.

A better way to do this would be to have it live with a web worker-specific tsconfig file, which would enable the `webworker` lib to provide proper context.

It would essentially be [this](https://stackoverflow.com/a/48950463/7617396), but extending the main config and just adding `webworker` in `lib`.

This should make `self` the proper type and give us type checking on everything.

Do note: the tsconfig file will apply to any file in that directory and sub-directories. Hence the need for workers to live alone, somewhere (Maybe as a sub-directory of `plugins` since they should be available globally, not sure)

File: tsconfig.json [hunk 5]
```diff
@@ -3,7 +3,12 @@
     "target": "ES2018",
     "module": "ESNext",
     "moduleResolution": "Node",
-    "lib": ["ESNext", "ESNext.AsyncIterable", "DOM"],
+    "lib": [
+      "ESNext",
+      "ESNext.AsyncIterable",
+      "DOM",
+      "webworker"
```
Comment (MrTimscampi) [inline tsconfig.json:10]: This should not be part of the main config, but live in its own sub-config specific to workers, since it collides with around 16 DOM types.

**Review by MrTimscampi — CHANGES_REQUESTED** [review]

**Comment by MrTimscampi** [comment]
Please also ensure that ESLint and Jest are happy :)

File: components/BlurhashCanvas.vue (hunk 1 above)
Comment (ferferga) [reply components/BlurhashCanvas.vue:8]: I checked with a simple ``console.log()`` in the worker and this doesn't happen.

The portion of the code that gets executed everytime a new blurhash canvas is invoked starts in ``export default Vue.extend(`` line. 

If it worked the way you mention there would be a lot of overhead not only because of the workers, but also because of the constant imports ^^

File: components/BlurhashCanvas.vue (hunk 3 above)
Comment (ferferga) [reply components/BlurhashCanvas.vue:49]: I've been all day and afternoon trying to do this and I couldn't. What I tried:

* Define a ``pixels`` data() prop => **Data prop seems to be shared across BlurhashCanvas instances, so consequent instances called after one instance would receive the results from the worker, applying the pixels of the first element in the row to the last one, for instance.
* Use a watcher => I either did it wrong or Vue seems to "fire and forget" about the canvas, so same thing happened as in the first attempt: we were placing pixels in the wrong cards.
* Multiple combinations I can't remember now.

Everything led me to the conclusion that ``this.$refs.canvas as HTMLCanvasElement`` points to the DOM node that it's currently being created, so the event handler will always lag behind, applying it to the incorrect node, leaving some cards without blurhash. It's strange though, because the ``onmessage`` function never reported having anything ``undefined`` and I tracked the process with logging everywhere and everything seemed to be right (except from the canvas that draw the pixels!).

**How I solved it?

It's ugly, but the only solution I found was to add the canvas node to a dict where the hash is the key. In the ``onmessage`` event, we receive the original hash and the pixels from the worker. With this, I get the canvas node and draw the pixels. Once it's done, I remove the element in the array.

I also complemented this by storing the decoded pixels in another array. Not sure if this is being caused by Virtual Scroller switching DOMs, but every time you scroll and see items that were not previously in the viewport, you'll see that they're generating blurhashes, which is unefficient. I didn't cover the fact that the array has no limits so it can eat up a lot of memory (wanted first to know your opinions in how to deal with this and how to solve this whole problem) but it seems to be very effective: Take in mind that the array where we "cache" the blurhashes is defined globally for the ``BlurhashCanvas`` component, like the worker, so every blurhash of every image will be hold on memory as soon as the page is reloaded. The memory footprint seems to be really low: I scrolled through my entire movie collection (~500 items), TV Shows (~100 items) and music (+1200 items) multiple times and the highest memory usage recorded by Chrome debugger was between **200-450 MB**, in a cycle (I guess the spikes were caused by Virtual Scroller recycling the DOM).

Comment (ferferga) [reply components/BlurhashCanvas.vue:39]: @MrTimscampi See [here](https://github.com/jellyfin/jellyfin-vue/pull/213/files#diff-fdffe0dcc16277333627cbefb4982d475d27eb2f237f92efb9376dda66d5a8ebL31-R57)

**Review by MrTimscampi — CHANGES_REQUESTED** [review]
Some really small things and a bigger concern.

With this, I don't get any blurhash in the app anymore. When scrolling, I'll sometimes get one to show up, but it seems very rare.

It seems to be a concurrency issue: the decoding doesn't block the main thread, so the main thread renders the image first, before any of the hashes come back. And indeed, if the main thread stalls, the blurhashes show up properly.

I'm not sure how to best handle this, right now, but I have a proposed solution for later.

We'll need to eventually move everything that has an item to Vuex stores, mainly to allow updating items globally when we receive websocket messages. As part of this, we can modify the items to pre-decode the hashes when adding them to the store. This ensures that every item has a pixel array available, which should fix the rendering issue.

This solution needs further large-scale refactoring though, so it's probably okay to merge this as-is (once the few small issues below are solved)

**Comment by MrTimscampi** [comment]
More thoughts on this:

Since we only currently support evergreen browsers, I think it'd be worth using [blurhash-rust-wasm](https://github.com/fpapado/blurhash-rust-wasm) in order to simplify the implementation and make decoding much faster.

Originally, that's what I used, but it was taken out to help a user test quickly and never brought back in due to lack of time. I think switching back to it would be best, since anything that can't run Web Assembly probably will probably be too slow for Blurhash anyway.

**Comment by ferferga** [comment]
@MrTimscampi What about keeping both and leaving the JS one as fallback? Workers already improve the JS situation and there might be devices (like Android 4 ones afaik, which are still really popular) that doesn't have WebAssembly (as it's still a new thing) but they're fine for blurhash.

**Comment by MrTimscampi** [comment]
> @MrTimscampi What about keeping both and leaving the JS one as fallback? Workers already improve the JS situation and there might be devices (like Android 4 ones afaik, which are still really popular) that doesn't have WebAssembly (as it's still a new thing) but they're fine for blurhash.

Fair enough, sounds good to me :smiley:

File: components/BlurhashCanvas.vue [hunk 6]
```diff
@@ -25,21 +25,14 @@ export default Vue.extend({
       default: 1
     }
   },
-  watch: {
-    hash() {
-      this.$nextTick(() => {
-        this.draw();
-      });
-    }
-  },
```
Comment (ferferga) [inline components/BlurhashCanvas.vue:34]: I'm not sure why we used this in the first place (in fact I don't remember removing it 😋), but everything is working fine on my end without it.

**Comment by ferferga** [comment]
Took a while but I'm excited that I got Comlink working the way it's doing right now. Workers are extremely easy to use, finally!

File: plugins/workers/blurhash.worker.ts [hunk 7]
```diff
@@ -0,0 +1,22 @@
...
+export default async function getPixels(
+  hash: string,
+  width: number,
+  height: number
+): Promise<Uint8ClampedArray> {
+  try {
+    const pixels = await decode(hash, width, height);
+    return pixels;
+  } catch {
+    throw new TypeError('Blurhash' + hash + ' is not valid');
```
Comment (Camc314) [inline plugins/workers/blurhash.worker.ts:20]:
```suggestion
    throw new TypeError(`Blurhash ${hash} is not valid`);

File: components/BlurhashCanvas.vue [hunk 8]
```diff
@@ -25,21 +25,14 @@ export default Vue.extend({
       default: 1
     }
   },
-  watch: {
```
Comment (Camc314) [inline components/BlurhashCanvas.vue:28]: Shouldnt we still have a `watch` method in case the hash changes

File: nuxt.config.ts [hunk 9]
```diff
@@ -228,6 +228,22 @@ const config: NuxtConfig = {
         ];
       }
     },
+    extend(config, { isDev, isClient }) {
```
Comment (Camc314) [inline nuxt.config.ts:231]: Linter's not happy

File: plugins/workers/blurhash.worker.ts [hunk 10]
```diff
@@ -0,0 +1,22 @@
...
+ * @param {height} height - Height of the decoded pixel array
+ * @returns {Promise} Returns the decoded pixels in the proxied response by Comlink
+ */
+export default async function getPixels(
+  hash: string,
+  width: number,
+  height: number
+): Promise<Uint8ClampedArray> {
+  try {
+    const pixels = await decode(hash, width, height);
```
Comment (Camc314) [inline plugins/workers/blurhash.worker.ts:17]: `decode` isn't async?

File: nuxt.config.ts (hunk 9 above)
Comment (ferferga) [reply nuxt.config.ts:231]: I have no idea how to fix this, it's been like this since day one, as Nuxt generates this file this way.

Maybe @MrTimscampi has an idea.

File: components/BlurhashCanvas.vue (hunk 8 above)
Comment (ferferga) [reply components/BlurhashCanvas.vue:28]: @Camc314 That's why I asked, but when?

Comment (ferferga) [reply components/BlurhashCanvas.vue:28]: @Camc314 That's why I asked, but when?

Comment (Camc314) [reply components/BlurhashCanvas.vue:28]: 🤷‍♀️ I guess in the case that a library filter is applied/removed, and a new card is not generated, but a new has is recieved

File: nuxt.config.ts (hunk 9 above)
Comment (Camc314) [reply nuxt.config.ts:231]: isDev can be removed
isClient is a bool

config is the webpack config object. Probably `any`. Not ideal though

Comment (Camc314) [reply nuxt.config.ts:231]:
```suggestion
    extend(config: any, { isClient }: { isClient: boolean }): void {
```

File: plugins/workers/blurhash.worker.ts (hunk 10 above)
Comment (ferferga) [reply plugins/workers/blurhash.worker.ts:17]: We don't need this. Thought at first that comlink wasn't working because the function needed to be async, but it was something else, however, forgot about removing this 😋.

TLDR: The call to the worker needs to be awaited, the worker function does not.

File: components/BlurhashCanvas.vue (hunk 8 above)
Comment (MrTimscampi) [reply components/BlurhashCanvas.vue:28]: It makes the canvas reactive to changes above it :)

DOM recycling, for example, just switches the properties of cards instead of destroying/recreating them. So anything that would lead to a change needs to react.

Here, if the item of the card changes, the hash changes, so the canvas would have to re-draw with the new pixel array.

Comment (MrTimscampi) [reply components/BlurhashCanvas.vue:28]: Also, once we have the websockets going, we'll get item updates notifications, so we can re-fetch a single item and update its card in place :yum:

Comment (ferferga) [reply components/BlurhashCanvas.vue]: @MrTimscampi Keeping this as it was originally.

File: components/BlurhashCanvas.vue [hunk 11]
```diff
@@ -36,16 +36,16 @@ export default Vue.extend({
...
+      const ctx = (this.$refs.canvas as HTMLCanvasElement).getContext('2d');
+      const imageData = ctx?.createImageData(this.width, this.height);
+      try {
+        const pixels = await getPixels(this.hash, this.width, this.height);
         if (imageData) {
           imageData.data.set(pixels);
           ctx?.putImageData(imageData, 0, 0);
         }
-      }
+      } catch {}
```
Comment (Camc314) [inline components/BlurhashCanvas.vue:48]: #306 Hides the canvas if the image is not valid

File: plugins/workers/blurhash.worker.ts [hunk 12]
```diff
@@ -0,0 +1,22 @@
...
+ * @returns {Promise} Returns the decoded pixels in the proxied response by Comlink
+ */
+export default function getPixels(
+  hash: string,
+  width: number,
+  height: number
+): Uint8ClampedArray {
+  try {
+    const pixels = decode(hash, width, height);
+    return pixels;
```
Comment (Camc314) [inline plugins/workers/blurhash.worker.ts:18]:
```suggestion
    return decode(hash, width, height);

File: components/BlurhashCanvas.vue (hunk 11 above)
Comment (ferferga) [reply components/BlurhashCanvas.vue]: @Camc134 What about setting validHash in the catch?

I want to keep the try-catch block so whatever error that could happen in the worker doesn't hang the component lifecycle and it's printed in the console. Detecting if the Blurhash is valid with isBlurhashValid is also more expensive than doing it this way (since a "partial" decode happens) and not always returns false when it should (I have one item that has a corrupt Blurhash due to a corrupted image, so I tested all of this).

Try-catch is as inexpensive as it could get in the main thread.

Comment (Camc314) [reply components/BlurhashCanvas.vue]: That's fine. I was just leaving the comment there as a note. When I update that PR, I'll update uit

**Review by Camc314 — APPROVED** [review]
LGTM

**Review by MrTimscampi — APPROVED** [review]

**Comment by ferferga** [comment]
Leaving this just for the record, as it only was mentioned in Matrix:

``blurhash-wasm`` wasn't added because I couldn't get it working, even in the main thread. This can be revisited later.
